Crypto/src/PKCS12Container.cpp

```diff
@@ -0,0 +1,140 @@
+//
+// PKCS12Container.cpp
+//
+// $Id: //poco/1.4/Crypto/src/PKCS12Container.cpp#1 $
+//
+// Library: Crypto
+// Package: Certificate
+// Module:  PKCS12Container
+//
+// Copyright (c) 2006-2009, Applied Informatics Software Engineering GmbH.
+// and Contributors.
+//
+// SPDX-License-Identifier:	BSL-1.0
+//
+
+
+#include "Poco/Crypto/PKCS12Container.h"
+#include "Poco/NumberFormatter.h"
+#include "Poco/StreamCopier.h"
+#include <sstream>
+#include <openssl/err.h>
+
+
+namespace Poco {
+namespace Crypto {
+
+
+PKCS12Container::PKCS12Container(std::istream& istr, const std::string& password)
+{
+	std::ostringstream ostr;
+	Poco::StreamCopier::copyStream(istr, ostr);
+	std::string cont = ostr.str();
+
+	BIO *pBIO = BIO_new_mem_buf(const_cast<char*>(cont.data()), static_cast<int>(cont.size()));
+	if (pBIO)
+	{
+		PKCS12* pkcs12 = d2i_PKCS12_bio(pBIO, NULL);
+		BIO_free(pBIO);
+		load(pkcs12, password);
+	}
+	else
+	{
+		throw NullPointerException("PKCS12Container BIO memory buffer");
+	}
+}
+
+
+PKCS12Container::PKCS12Container(const Poco::Path& path, const std::string& password)
+{
+	FILE* pFile = nullptr;
+
+	if ((pFile = fopen(path.toString().c_str(), "rb")))
+	{
+		PKCS12* pPKCS12 = d2i_PKCS12_fp(pFile, NULL);
+		fclose (pFile);
+		load(pPKCS12, password);
+	}
+	else
+	{
+		throw NullPointerException("PKCS12Container file:" + path.toString());
+	}
+}
+
+
+PKCS12Container::PKCS12Container(const std::string& str, const std::string& password)
+{
+}
+
+
+PKCS12Container::PKCS12Container(const PKCS12Container& cont)
+{
+}
+
+
+PKCS12Container& PKCS12Container::operator = (const PKCS12Container& cert)
+{
+	return *this;
+}
+
+
+PKCS12Container::~PKCS12Container()
+{
+	if (_pKey) EVP_PKEY_free(_pKey);
+}
+
+
+void PKCS12Container::load(PKCS12* pPKCS12, const std::string& password)
+{
+	if (pPKCS12)
+	{
+		X509* pCert = nullptr;
+		STACK_OF(X509)* pCA = nullptr;
+		STACK_OF(PKCS12_SAFEBAG)* bags = nullptr;
+		if (PKCS12_parse(pPKCS12, password.c_str(), &_pKey, &pCert, &pCA))
+		{
+			if (pCert && _pKey)
```

Comment:
I think it is possible to hava a PKCS12 container without a key. (See: `trusted.p12`)

I created some test files: [ca.zip](https://github.com/pocoproject/poco/files/1291993/ca.zip)


`server.p12` containing:
- server.key - private key
- server.pem - certificate
- ca-chain.pem - certificate chain with root and intermediate
created: `openssl pkcs12 -export -out server.p12 -inkey intermediate/private/server.key.pem -in intermediate/certs/server.cert.pem -certfile intermediate/certs/ca-chain.cert.pem -name vally`
password: `crypto`
info: `openssl pkcs12 -info -in server.p12`


`trusted.p12` containing:
- list of certificates
created: `openssl pkcs12 -export -out trusted.p12 -in intermediate/certs/ca-chain.cert.pem -certfile trusted-certs.pem -nokeys -caname vally-ca -name vally -caname vally-ca`
password: `crypto`
info: `openssl pkcs12 -info -in trusted.p12`

The files I created can be used for unit tests in poco if desired

Reply:
these were helpful, thanks